lib/perl/Genome/Model/Tools/Picard/Base.pm

```diff
@@ -0,0 +1,169 @@
+}
+
+# given a property meta object (defining property name, and ostensibly
+# picard_param_name), return the list of cmdline args that should be passed
+# to picard based on the values set on the current object.
+#
+# this is where is_many properties get handled.
+sub _picard_args_from_meta {
+    my ($self, $meta) = @_;
+
+    my $ur_name = $meta->property_name;
+    # using // instead of ? : below made a lims perl test fail
+    my $type = defined $meta->data_type ? $meta->data_type : 'String';
+    my $picard_name = $meta->picard_param_name;
+
+    return unless defined $self->$ur_name;
+
+    # This works for things that are is_many or not.
+    return map {_format_picard_arg($type, $picard_name, $_)} $self->$ur_name;
+}
+
+# command line args that come after the jar and class
+#
+# The default implementation just translates all the inputs that have
+# picard_param_name defined. Boolean values get translated to true/false,
+# and is many attributes get repeated, e.g.,
+#
+#  create_md5_file => {..., is => 'Boolean'}
+#       when set to something that evalutes to false
+#       yields
+#  qw(CREATE_MD5_FILE=false)
+#
+# and
+#
+#  program_name => {..., is_many => 1}
+#       when set to ['a', 'b', 'c']
+#       yields
+#  qw(PROGRAM_NAME=a PROGRAM_NAME=b PROGRAM_NAME=c)
+#
+# Override for custom behavior.
+sub _cmdline_args {
+    my $self = shift;
+
+    # CREATE_MD5_FILE will cause errors for certain picard commands (e.g., CompareSAMs)
+    # the picard default value is false.
+    my @metas = grep {$_->property_name ne 'create_md5_file'} $self->_picard_param_metas;
+    my @args = map {$self->_picard_args_from_meta($_)} @metas;
+
+    push @args, 'CREATE_MD5_FILE=true' if $self->create_md5_file;
+    return @args;
+}
+
+# this is your chance to make a loud sound and die if a caller asks for
+# something ridiculous. params can also be edited here if that is somehow
+# appropriate.
+sub _validate_params {
+    # example:
+    # my $self = shift;
+    # $self->enforce_minimum_version('1.85');
+    # die unless -s $self->input_file;
+    # $self->be_noisy(0) unless $self->log_file;
+}
+
+# want to pass extra stuff to Genome::Sys->shellcmd? return a hash (not ref)
+# of those params here.
+sub _shellcmd_extra_params {
+    # example:
+    # my $self = shift
+    # return (
+    #   input_files => [$self->input_file, $self->reference_sequence],
+    #   output_files => [$self->output_file]
+    #   );
+}
+
+# string to be appended after the full command line (e.g., '> foo.out')
+#
+# there was a bug that led to some commands being built like:
+#
+#   java ... foo.jar class ARGS > output_file MORE_ARGS
+#
+# that's bad. this sort of conflicts with the idea of log_file, but commands
+# like "BamIndexStats" that have no native OUTPUT option (i.e., can only spew
+# to stdout) support the existence of this function.
+sub _redirects {
+    my $self = shift;
+    return sprintf(">> %s", $self->log_file) if $self->log_file;
```

Comment:
Yeah, that extra newline is a bummer.  It stopped me from using `redirect_stdout` recently.  The name `_redirects` doesn't seem so bad, since that's what it's currently used for.